### ISSUE TITLE
Argument Null Exception on Parse/ ParseAsync

### ISSUE THREAD

**Issue by Manthazar** [Http-Multipart-Data-Parser/Http-Multipart-Data-Parser#104]
I try to work around the lack of form data of Azure Function Http Trigger. Following examples, I have a following helper method which fails for some reason:

```
   public static async Task<T> GetFormData<T>(this HttpRequestData request, string parameterName)
  {
            Guard.ArgumentNotNull(nameof(request), request);
            Guard.ArgumentNotNullOrEmpty(nameof(parameterName), parameterName);

            var parser = await MultipartFormDataParser.ParseAsync(request.Body);
            var jsonValue = parser.GetParameterValue(parameterName);
            var jsonSettings = ConfigurationModule.ConfigureNewtonsofJsontOptions(new JsonSerializerSettings());
            var result = JsonConvert.DeserializeObject<T>(jsonValue, jsonSettings);

            return result;
        }
```

I am getting the error 'System.ArgumentNullException: 'Value cannot be null. (Parameter 'source')' at the line `var parser = await MultipartFormDataParser.ParseAsync(request.Body);`

Package References:
```
PackageReference Include="Microsoft.Azure.Functions.Worker" Version="1.6.0" 
PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.Http" Version="3.0.12" 
PackageReference Include="HttpMultipartParser" Version="5.0.1"
```

![image](https://user-images.githubusercontent.com/76886475/156603332-0da6193f-e8f7-412b-8fe5-7689c26fc71e.png)

**Comment by Manthazar** [comment]
Update: Found that the reason is actually related to my attempts to get mitigate the functional gap on the Azure Function Http Trigger: I had added a `var requestString = await requestData.ReadAsStringAsync();` call right before calling this method.

So the issue I call out is actually  about a **bad error message** only.: it should inform that something is fishy with the input stream rather than complaining about an inexistent argument.

This should reproduce the exception:
 public static async Task<T> GetFormData<T>(this HttpRequestData request, string parameterName)
  {
            Guard.ArgumentNotNull(nameof(request), request);
            Guard.ArgumentNotNullOrEmpty(nameof(parameterName), parameterName);

            var requestString = await requestData.ReadAsStringAsync(); // does something with the body stream so that the parser collapses

            var parser = await MultipartFormDataParser.ParseAsync(request.Body);
            var jsonValue = parser.GetParameterValue(parameterName);
            var jsonSettings = ConfigurationModule.ConfigureNewtonsofJsontOptions(new JsonSerializerSettings());
            var result = JsonConvert.DeserializeObject<T>(jsonValue, jsonSettings);

            return result;
        }

**Comment by Jericho** [comment]
What you are describing is discussed [here](https://github.com/Http-Multipart-Data-Parser/Http-Multipart-Data-Parser/issues/66).

The root cause of the problem is: when you read the content of a stream, you end up moving the position within the stream. In your example, you read the entirety of the stream content with `ReadAsStringAsync` which bring the position to the very end of the stream. When you subsequently attempt to parse the content of the stream, there's no content left to be read. 

A quick solution would be to "rewind" the stream before attempting to parse the content with something like: `myStream.Position = 0;`. In you case, the stream in question is the body of the request which may or may not be "seekable" therefore it's possible you may get an exception when you try to set the position (I believe that different versions of ASP.NET have different rules regarding whether a HTTP request content stream is seekable or not, that's why I say that you "may" get an exception).

Having said all this, I agree that the exception is not helpful. I'll see if I can improve it.

**Comment by Jericho** [comment]
By the way, what do you suggest would be a better error message?

My first suggestion is: `The content does not appear to be valid: it's either malformed or empty.` which is concise but I'm afraid it doesn't help developers figuring out how to resolve the problem.

My second suggestion is: `The content does not appear to be valid: it's either malformed or empty. The most likely cause is that you have read the content prior to attempting to parse it which has moved the 'Position' within the stream to the end and resulted in no more content to be read. If that's the case, you need to rewind the stream prior to attempting to parse its content.` which is very descriptive and proposes a solution but it might be too verbose.

I'm opened to suggestions.

**Comment by Manthazar** [comment]
Hm, it's easier to complain ;) I think -in my case above- if the tool would have been throwing a System.IO.EndOfStreamException, my search (for my error) would have been much quicker. Maybe, this could be detected by checking Position/ Length of the stream, i.e. when they are equal.

Thinking about the case when the given stream is not at its end, but the content is fishy, I would join your first option as a FormatException. In cases which originate in frequent errors/ mistakes, my rather personal style could make me to add a question to the error message. 

Summarized, my tendency would be:
```
- case given stream position and length > 0 and equal each other --> System.IO.EndOfStreamException
- case given stream length =0 -->  System.FormatException with message 'The content does not appear to be valid: empty.'
- case given stream length >0 and position >0 and parsing error --> System.FormatException: . `The content does not appear to be valid. Did you intend to start parsing at position <xyz>?
```

Not sure though that the last one is easy to do. No need to over-engineer :)

**Comment by Jericho** [comment]
Thanks for sharing your thoughts. Unfortunately, I can't rely on being able to know the "position" in the stream because not all classes that derive from the `Stream` base class allow reading this value. However, I was able to narrow down the problematic situations to two specific scenarios which allows me to throw a meaningful exception for each one:

- **Scenario 1:**
The parser is invoked with either an empty stream or the 'Position' in the stream has been moved to the end. This can be reproduced with the following code:
```csharp
var requestContent = await Request.ReadAsStringAsync(); // Reading the content of the stream causes the Position to move to the end of the stream
var parser = await MultipartFormDataParser.ParseAsync(Request.Body);
```

The consequence is that we get a null value when we read the content of the stream in order to determine what the multi-part boundary is. As far as the parser is concerned, this is not a valid situation and we must throw an exception. Currently the exception is not helpful: `Value cannot be null. (Parameter 'source')`. I am solving this scenario by throwing a more descriptive exception:
```csharp
throw new MultipartParseException("Unable to determine boundary: unexpected end of stream");
```

This scenario roughly corresponds to your first two suggestions.

- **Scenario 2:**
The parser is invoked with the 'Position' in the stream moved to an arbitrary position. This can be reproduced with the following code:
```csharp
Request.Body.Position = 25; // Move to an arbitrary position
var parser = await MultipartFormDataParser.ParseAsync(Request.Body).ConfigureAwait(false);
```

The consequence is that we get malformed data when we read the content of the stream, which leads us to incorrectly determine the multi-part boundary and therefore all the remaining content is ignored due to the fact that it is not separated by the incorrect boundary. Ultimately, the parser throws a `Unexpected end of stream` exception which is confusing. I improved this scenario by throwing a more meaningful exception:
```csharp
throw new MultipartParseException("Unable to determine boundary: content does not start with a valid multipart boundary");
```

This scenario roughly corresponds to your third suggestion.
